HPXMLtoOpenStudio/tests/test_validation.rb

```diff
@@ -0,0 +1,264 @@
+# frozen_string_literal: true
+
+require_relative '../resources/minitest_helper'
+require 'openstudio'
+require 'openstudio/ruleset/ShowRunnerOutput'
+require 'minitest/autorun'
+require 'fileutils'
+require_relative '../measure.rb'
+require_relative '../resources/util.rb'
+require 'schematron-nokogiri'
+
+class HPXMLtoOpenStudioValidationTest < MiniTest::Test
+  def before_setup
+    @root_path = File.absolute_path(File.join(File.dirname(__FILE__), '..', '..'))
+
+    # load the Schematron xml
+    @stron_path = File.join(@root_path, 'HPXMLtoOpenStudio', 'resources', 'EPvalidator.xml')
+    # make a Schematron object
+    @stron_doc = SchematronNokogiri::Schema.new Nokogiri::XML File.open(@stron_path)
+
+    # Load all HPXMLs
+    @hpxml_docs = {}
+    sample_files_dir = File.absolute_path(File.join(@root_path, 'workflow', 'sample_files'))
+    Dir["#{sample_files_dir}/*.xml"].sort.each do |xml|
+      @hpxml_docs[File.basename(xml)] = HPXML.new(hpxml_path: File.join(sample_files_dir, File.basename(xml))).to_oga()
+    end
+  end
+
+  def test_sample_files
+    puts "Testing #{@hpxml_docs.size} HPXML files..."
+    @hpxml_docs.each do |xml, hpxml_doc|
+      print '.'
+
+      # HPXML Schema validation
+      _test_schema_validation(hpxml_doc, xml)
+      # Ruby validator validation
+      _test_ruby_validation(hpxml_doc)
+      # Schematron validation
+      _test_schematron_validation(@stron_doc, hpxml_doc.to_xml)
+    end
+
+    puts
+  end
+
+  def test_schematron_asserts
+    # build up expected error messages hashes by parsing EPvalidator.xml
+    doc = XMLHelper.parse_file(@stron_path)
+    expected_error_msgs_by_element_addition = {}
+    expected_error_msgs_by_element_deletion = {}
+    XMLHelper.get_elements(doc, '/sch:schema/sch:pattern').each do |pattern|
+      XMLHelper.get_elements(pattern, 'sch:rule').each do |rule|
+        rule_context = XMLHelper.get_attribute_value(rule, 'context')
+
+        XMLHelper.get_values(rule, 'sch:assert').each do |assertion|
+          parent_xpath = rule_context.gsub('h:', '').gsub('/*', '')
+          element_name = _get_element_name(assertion)
+          target_xpath = [parent_xpath, element_name]
+          expected_error_message = _get_expected_error_message(parent_xpath, assertion)
+
+          if assertion.start_with?('Expected 0') || assertion.partition(': ').last.start_with?('[not') # FIXME: Is there another way to do this?
+            next if assertion.start_with?('Expected 0 or more') # no tests needed
+
+            expected_error_msgs_by_element_addition[target_xpath] = expected_error_message
+          elsif assertion.start_with?('Expected 1') || assertion.start_with?('Expected 9')
+            expected_error_msgs_by_element_deletion[target_xpath] = expected_error_message
```

Comment:
I don't understand why it's either element addition or deletion but never both. If we have an "Expected 1" situation, for example, shouldn't we check that it fails both element addition _and_ deletion?

Reply:
If I remember correctly, we talked about this before. I think you're right. We should check if all assertion passes or fails by both element addition and deletion. Let me address this.